index.html

```diff
@@ -1,390 +1,1451 @@
+      failures by opening a new reader. Fatal errors will cause
+      {{SerialPort/readable}} to become `null` and the loop to end.
+
+      <p>
+      As long as the serial port is open it can continue to produce data and the
+      amount of data in each of the chunks returned by
+      {{ReadableStreamDefaultReader/read()}} will be essentially arbitrary based
+      on the timing of when it is called. It is up to the device and the code
+      communicating with it to decide what constitutes a complete message. For
+      example, a device might communicate with the host using ASCII-formatted
+      text where each message ends with a newline (or the sequence `"\r\n"`). A
+      pipeline of {{TransformStream}}s can be used to automatically convert the
+      {{Uint8Array}} chunks provided by {{SerialPort/readable}} into
+      {{DOMString}}s containing an entire line of text each.
+
+      <pre class="js">
+        class LineBreakTransformer {
+          constructor() {
+            this.container = '';
+          }
+
+          transform(chunk, controller) {
+            this.container += chunk;
+            const lines = this.container.split('\r\n');
+            this.container = lines.pop();
+            lines.forEach(line => controller.enqueue(line));
+          }
+
+          flush(controller) {
+            controller.enqueue(this.container);
+          }
+        }
+
+        const lineReader = port.readable
+            .pipeThrough(new TextDecoderStream())
+            .pipeThrough(new TransformStream(new LineBreakTransformer()))
+            .getReader();
+      </pre>
+
+      Some other ways of encoding message boundaries are to prefix each message
+      with its length or to wait a defined length of time before transmitting
+      the next message. Implementing a {{TransformStream}} for these types of
+      message boundaries is left as an exercise for the reader.
+    </aside>
+
+    The {{SerialPort/readable}} getter steps are:
+
+    <ol>
+      <li>
+        If [=this=].{{[[readable]]}} is not `null`, return
+        [=this=].{{[[readable]]}}.
+      <li>If [=this=].{{[[state]]}} is not `"opened"`, return `null`.
+      <li>If [=this=].{{[[readFatal]]}} is `true`, return `null`.
+      <li>Let |stream| be a [=new=] {{ReadableStream}}.
+      <li>
+        Let |pullAlgorithm| be the following steps, given
+        |controller:ReadableStreamDefaultController|:
+        <ol>
+          <li>Let |promise:Promise| be [=a new promise=].
+          <li>[=Resolve=] |promise| with `undefined`.
+          <li>
+            Return |promise| and run the following steps [=in parallel=]:
+            <ol>
+              <li>
+                Let |desiredSize| be
+                |controller|.{{ReadableStreamDefaultController/desiredSize}}
+              <li>
+                Let |buffer| be a newly created {{ArrayBuffer}} of |desiredSize|
+                bytes.
```

Comment:
This strategy of always creating an ArrayBuffer of size desiredSize, then creating a Uint8Array view spanning potentially less than that, prohibits smarter strategies of only allocating enough memory for the amount of data read.

It might be better to do something like:

- Read up to _desiredSize_ bytes, putting the result in the [byte sequence](https://infra.spec.whatwg.org/#byte-sequence) _bytes_
- Create an ArrayBuffer _buffer_ from bytes, optionally padded with extra 0s so that its length is between _bytes_'s length and _desiredSize_.
- Create a Uint8Array view onto _buffer_, whose length is _bytes_'s length.

Reply:
Something I don't think I've done is specify how `bufferSize` turns into the `desiredSize` on the `ReadableStream`. Similar to the algorithms for the `WritableStream` in Blink there is actually a shared memory buffer of `bufferSize` bytes the browser process is constantly trying to fill by reading from the port and the HWM is set to 1 in the stream itself so that it doesn't cause any additional buffering.

Reply:
Oh, right, good point! Right now the desiredSize is probably always just 1... There are a few ways you could do this, but I'm trying to think of the one that strikes the best balance between simplicity of specification, implementation flexibility, and interoperability.

I think the one that makes the most sense to me is to create the stream with an implementation-defined high water mark (your `bufferSize`), and a byte-counting size algorithm (*). Then you can create the chunk as you do here, or with the slight modification I suggested.

Note that with this architecture, generally `desiredSize` will be: implementation defined high water mark, minus the size of any queued-up-but-not-yet-read chunks. Recall that the streams machinery will call your pullAlgorithm whenever the total size of queued-up-but-not-yet-read chunks is less than the high water mark.

This communicates to the spec reader that implementations have some size they're generally targeting to buffer up. But that size is implementation-defined. And strategies for filling it could vary; this doesn't spell out the shared-memory strategy. (Although doing so in a note might be nice.)

Alternately, none of these details is really observable from JavaScript, so you could go further and just have the pullAlgorithm read an implementation-defined amount of bytes. This loses the communication that implementations generally aim for consistency, but is a good bit simpler.

---

(*): The Streams spec should define "byte-counting size algorithm" for you. It currently does not. Feel free to leave it unlinked for now and I can add it, once we're happy with this text.